src/utils/deduper.zig

```diff
@@ -0,0 +1,203 @@
+            var bits = try std.ArrayList(AtomicU64).initCapacity(allocator, size);
+            for (0..size) |_| bits.appendAssumeCapacity(AtomicU64.init(0));
+            var state: [n_hashers]AHashSeed = undefined;
+            for (0..n_hashers) |i| state[i] = AHashSeed.fromRng(rand);
+            return .{
+                .num_bits = num_bits,
+                .bits = bits,
+                .state = state,
+                .last_reset_instant = Instant.now(),
+                .masked_count = AtomicU64.init(0),
+            };
+        }
+
+        pub fn deinit(self: Deduper(n_hashers, T)) void {
+            self.bits.deinit();
+        }
+
+        /// Resets the deduper if the false positive rate is too high or the reset cycle has elapsed.
+        pub fn maybeReset(
+            self: *Deduper(n_hashers, T),
+            rand: std.rand.Random,
+            false_positive_rate: f64,
+            reset_cycle: Duration,
+        ) bool {
+            std.debug.assert(0.0 < false_positive_rate and false_positive_rate < 1.0);
+            const saturated = self.falsePositiveRate() >= false_positive_rate;
+            if (saturated or
+                self.last_reset_instant.elapsed().asNanos() >= reset_cycle.asNanos())
+            {
+                for (self.bits.items) |_bit| {
+                    var bit = _bit;
+                    bit = AtomicU64.init(0);
+                }
+                for (0..n_hashers) |i| self.state[i] = AHashSeed.fromRng(rand);
+                self.last_reset_instant = Instant.now();
+                self.masked_count = AtomicU64.init(0);
+            }
+            return saturated;
+        }
+
+        /// Returns true if the data is likely a duplicate.
+        pub fn dedup(self: *Deduper(n_hashers, T), data: *const T) bool {
+            var duplicate = true;
+            for (0..n_hashers) |i| {
+                var hasher = AHasher.fromSeed(self.state[i]);
+                hasher.hash(T, data);
+                const hash: u64 = hasher.finish() % self.num_bits;
+                const mask: u64 = @as(u64, 1) << @truncate(hash);
+                const index = @as(usize, hash >> 6);
+                const old = self.bits.items[index].fetchOr(mask, .acquire);
+                if (old & mask == 0) {
+                    _ = self.masked_count.fetchAdd(1, .acquire);
+                    duplicate = false;
+                }
+            }
+            return duplicate;
+        }
+
+        /// False positive rate computed from the current popcount and num_bits.
+        pub fn falsePositiveRate(self: *Deduper(n_hashers, T)) f64 {
+            const popcount = self.masked_count.load(.unordered);
+            const numerator: f64 = @floatFromInt(@min(popcount, self.num_bits));
+            const denominator: f64 = @floatFromInt(self.num_bits);
+            const ones_ratio = numerator / denominator;
+            return std.math.pow(f64, ones_ratio, @as(f64, n_hashers));
```

Comment:
How did you come up with this formula? Intuitively it seems reasonable, but it's hard for me to prove that it's correct.

Reply:
Again, based off of the agave deduper. I think it will be best addressed in a follow up PR addressing your performance suggestions. We may even want to go for an entirely different deduper implementation, it was just the fastest and simplest route to demonstrating equivalent retransmision